scimma_admin/hopskotch_auth/models.py

```diff
@@ -202,3 +207,312 @@ def rand_username(owner: User) -> str:
     owner_emailname = owner.email.split("@")[0]
     rand_username_suffix = secrets.token_hex(nbytes=4)
     return f"{owner_emailname}-{rand_username_suffix}"
+
+
+class MembershipStatus(enum.Enum):
+    Member = 1
+    Owner = 2
+
+
+class Group(models.Model):
+    name = models.CharField(
+        max_length=256,
+        unique=True,
+    )
+
+
+class GroupMembership(models.Model):
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.CASCADE,
+    )
+    group = models.ForeignKey(
+        Group,
+        on_delete=models.CASCADE,
+    )
+    status = enum.EnumField(
+        MembershipStatus,
+        editable=True,
+    )
+
+
+def is_group_member(user, group):
+    try:
+        membership = GroupMembership.objects.get(user_id=user, group_id=group)
+        return membership.status == MembershipStatus.Member or membership.status == MembershipStatus.Owner
+    except ObjectDoesNotExist as dne:
+        # if there is no record the user is not in the group
+        return False
+
+
+def is_group_owner(user, group):
+    try:
+        membership = GroupMembership.objects.get(user_id=user, group_id=group)
+        return membership.status == MembershipStatus.Owner
+    except ObjectDoesNotExist as dne:
+        # if there is no record the user is not in the group and cannot be an owner
+        return False
+
+
+class KafkaTopic(models.Model):
+    owning_group = models.ForeignKey(
+        Group,
+        on_delete=models.CASCADE,
+    )
+    name = models.CharField(
+        max_length=249, # see https://github.com/apache/kafka/commit/ad3dfc6ab25c3f80d2425e24e72ae732b850dc60
+        editable=False,
+    )
+    publicly_readable = models.BooleanField(
+        default = False,
+    )
+
+
+def validate_topic_name(name: str) -> bool:
+    # https://github.com/apache/kafka/blob/bc55f85237cb46e73c6774298cf308060a4a739c/clients/src/main/java/org/apache/kafka/common/internals/Topic.java#L30
+    valid = re.compile("^[a-zA-Z0-9._-]{1,249}$")
+    return re.match(valid, name)
+
+
+class KafkaOperation(enum.Enum):
+    All = 1
+    Read = 2
+    Write = 3
+    Create = 4
+    Delete = 5
+    Alter = 6
+    Describe = 7
+    ClusterAction = 8
```

Comment:
I'm not sure; at the moment I have erred on the side of including every possible operation in case we need it later, but I could certainly see not displaying most of them in the UI.